Set sender/receiver capabilities to attach frame in the amqp10_client

MarcialRosales opened the pull request:
## Proposed Changes

It partially addresses the feature request https://github.com/rabbitmq/rabbitmq-server/issues/10752. This PR lets a developer set the capabilities to the sender and receiver links in the amqp10_client. It will not add this capability to the shovel plugin though.

The system test group `ibmmq` requires a docker image for IBM MQ server. This docker image is built by a workflow added by this other PR https://github.com/rabbitmq/rabbitmq-server/pull/11419. The official IBM MQ docker image does not come with AMQP built-in capability hence we have to build one. And given the amount of time it takes to build this image it was better to build it from its own CI workflow. 

NOTE for reviewer:
- To run system test group `ibmmq`, it pulls a docker image based on amd64 arch. It does not take into account the local arch because the other possible architecture is arm64 and it is not possible to build a docker image for ARM64 with AMQP capabilities. (see https://github.com/rabbitmq/rabbitmq-server/blob/set-amqp10-capabilities/deps/amqp10_client/test/system_SUITE_data/ibmmq_runner#L11 for more context)

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

ansd commented on deps/amqp10_client/src/amqp10_client_session.erl:
```diff
@@ -771,6 +779,19 @@ filter_value_type({T, _} = V) when is_atom(T) ->
     %% looks like an already tagged type, just pass it through
     V.
 
+translate_terminus_capabilities(none) ->
+    undefined;
+translate_terminus_capabilities(Capabilities) when is_binary(Capabilities) ->
+    {utf8, Capabilities};
+translate_terminus_capabilities(CapabilitiesList) when is_list(CapabilitiesList) ->
+    {list, [filter_capability(V) || V <- CapabilitiesList]}.
+
+filter_capability(V) when is_binary(V) ->
+    {utf8, V};
+filter_capability({T, _} = V) when is_atom(T) ->
+    %% looks like an already tagged type, just pass it through
+    V.
+        
```
Wrong AMQP types are used:
1. Capabilities are symbols, not strings.
2. Multiple values of a composite type are represented by an array, not by a list.

MarcialRosales replied:
It is encoded as an array and as a symbol. See the tests https://github.com/rabbitmq/rabbitmq-server/blob/set-amqp10-capabilities/deps/amqp10_client/test/system_SUITE.erl#L935 where we send an array of symbols. 

The API exposes or accepts Erlang types from users where it accepts binaries (strings) and list of binaries but the library is serliazing/deserlizing it as an array of symbols .. It is more or less how we treat Filters ...

MarcialRosales replied:
It is true that multiple values are encoded using `array` rather than `list`.

ansd reviewed (changes requested):
I suggest that we first test end-to-end that the RabbitMQ <-> IBM MQ integration works before merging this PR.
Right now, this PR adds a lot of boilerplate code that is not used and contains bugs.

MarcialRosales commented on deps/amqp10_client/src/amqp10_client.erl:
```diff
@@ -20,14 +20,17 @@
          attach_sender_link/3,
          attach_sender_link/4,
          attach_sender_link/5,
+         attach_sender_link/6,
          attach_sender_link_sync/3,
          attach_sender_link_sync/4,
          attach_sender_link_sync/5,
+         attach_sender_link_sync/6,
          attach_receiver_link/3,
          attach_receiver_link/4,
          attach_receiver_link/5,
          attach_receiver_link/6,
          attach_receiver_link/7,
+         attach_receiver_link/8,
          attach_link/2,
```
I will look into using `attach_link`. When I was doing it I thought there could be another way of doing this like setting a structure like Attach_args and passing it . But the API already presented this pattern or design . I just followed it.

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -157,9 +177,11 @@ end_per_group(_, Config) ->
 %% -------------------------------------------------------------------
 
 init_per_testcase(_Test, Config) ->
+    ct:log("Setting per test case"),
```
These log statements seem to be useless. Can we remove them?

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -157,9 +177,11 @@ end_per_group(_, Config) ->
     case lists:keyfind(mock_port, 1, Config) of
         {_, Port} ->
             M = mock_server:start(Port),
+            ct:log("Setting mock server"),
```
These log statements seem to be useless. Can we remove them?

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -176,19 +198,20 @@ open_close_connection(Config) ->
+        {amqp10_event, {connection, Connection2, opened}} -> ct:log("connection opened"), ok
     after 5000 -> exit(connection_timeout)
     end,
+    ct:log("Closing connection ..."),
```
These log statements seem to be useless. Can we remove them?

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -324,61 +347,107 @@ roundtrip_large_messages(Config) ->
+    ok = roundtrip(OpenConf, [{body, Data64Mb}]).
+
+basic_roundtrip_ibmmq(Config) ->
+    application:start(sasl),
```
Why do we need to start the Erlang `sasl` application here?
Is https://www.erlang.org/doc/apps/sasl/error_logging.html being confused with https://datatracker.ietf.org/doc/html/rfc4422 ?

ansd replied:
I ran the test without this line, and the test succeeds.

MarcialRosales replied:
Indeed . In fact, it is not needed in the function `basic_roundtrip` either. I removed it from there and the tests succeeds.

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -324,61 +347,107 @@ roundtrip_large_messages(Config) ->
+                   properties => #{}
+                   },
+    {ok, Sender} = amqp10_client:attach_link(Session, SenderAttachArgs),
+    %%await_link(Sender, credited, link_credit_timeout),
```
Delete commented code.

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -324,61 +347,107 @@ roundtrip_large_messages(Config) ->
     % ct:pal(?LOW_IMPORTANCE, "roundtrip message Out: ~tp~nIn: ~tp~n", [OutMsg, Msg]),
-    ?assertMatch(Props, amqp10_msg:properties(OutMsg)),
+    ActualProps = amqp10_msg:properties(OutMsg),
+    [ ?assertEqual(V, maps:get(K, ActualProps)) || K := V <- Props, K =/= creation_time],
```
The creation-time must not be changed by the broker as it's part of the properties which are part of the bare message and therefore immutable. If IBM MQ changes this creation time, this is a bug in IBM MQ. Can we maintain the assertion that creation time matches, and only omit this assertion for the buggy IBM MQ?

ansd commented on deps/amqp10_client/test/system_SUITE.erl:
```diff
@@ -324,61 +347,107 @@ roundtrip_large_messages(Config) ->
+
+filtered_roundtrip(OpenConf, Args) ->
+    Body = proplists:get_value(body, Args, <<"banana">>),
+    Destination = proplists:get_value(destination, Args, <<"test1">>),
```
Can we simplify this again as done prior to this PR?
Here, body is always `banana` and Destination is always `test1`.
There is no need to use `proplists:get_value/3` here.

ansd reviewed (changes requested):
I get sometimes the following error

```
system_SUITE > ibmmq > basic_roundtrip_ibmmq
    #1. {error,
            {shutdown,
                {gen_statem,call,
                    [<0.540.0>,
                     {attach,
                         #{name => <<"banana-sender">>,filter => #{},
                           properties => #{},rcv_settle_mode => first,
                           role =>
                               {sender,
                                   #{address => <<"DEV.QUEUE.3">>,
                                     capabilities => <<"queue">>,
                                     durable => unsettled_state}},
                           snd_settle_mode => settled}},
                     5000]}}}
```

when running

```
make -C deps/amqp10_client/ ct-system
```
locally on Ubuntu.
Might be worth checking this flake?

MarcialRosales commented on deps/amqp10_client/src/amqp10_client_session.erl:
```diff
@@ -73,11 +73,14 @@
 -type rcv_settle_mode() :: first | second.
 
 -type terminus_durability() :: none | configuration | unsettled_state.
+-type terminus_capabilities() :: none | binary() | list().
```
nice refinement ! Thanks

MarcialRosales committed:
Configure capabilities on the source/target field in the ATTACH frame

MarcialRosales committed:
Minor cleanup

michaelklishin commented:
The forced push was a rebase.

ansd commented:
The build currently fails because IBM MQ can't be started:
```
Error: {error,1,
              "Docker image pivotalrabbitmq/ibm-mqadvanced-server-dev:9.3.5.1-amd64 ready\nUnable to find image 'pivotalrabbitmq/ibm-mqadvanced-server-dev:9.3.5.1-amd64' locally\ndocker: Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.
```

ansd commented on deps/amqp10_client/src/amqp10_client_session.erl:
```diff
@@ -771,6 +778,17 @@ filter_value_type({T, _} = V) when is_atom(T) ->
+    {symbol, V};
+filter_capability(_) ->
+    %% Any other type is ignored
+    {}.
```
Is this code tested?
I'm pretty sure this tuple with 0 elements will crash later in `amqp10_binary_generator`.
See my comment in https://github.com/rabbitmq/rabbitmq-server/pull/11337#discussion_r1647195471

MarcialRosales committed:
Ignore sender/receiver non-binary capabilities

MarcialRosales committed:
interim change ct:log to be able to

log what capabilites are sent on each
test case scenario
it looks like some times the test case passes
because the non-binary value `dummy` is properly
skipped but sometimes the test cases fails because
it is not skipping it . i.e. is_binary(dummy) returns true.

MarcialRosales committed:
Use logger:debug to log capabilities

MarcialRosales committed:
Log with warning level instead

MarcialRosales committed:
Capture error and log it

To help troublehshoot an issue originated from
translate_terminus_capabilities

MarcialRosales committed:
Add missing log statement

MarcialRosales committed:
Change message

MarcialRosales committed:
Add extra logging on attach

MarcialRosales committed:
Add extra logging

MarcialRosales committed:
Add extra logging

MarcialRosales committed:
Fix issue

MarcialRosales committed:
Modified log statement

MarcialRosales committed:
REvert last change

MarcialRosales committed:
Log return from send_attach

MarcialRosales committed:
Test capabilities with activemq

MarcialRosales committed:
Log around sending frame

MarcialRosales committed:
Log eveny operation related to stop event

MarcialRosales committed:
Test capabilities with activemq and ibmmq

MarcialRosales committed:
Sleep for 500msec before considering ibmmq running

MarcialRosales committed:
Fix issue in the ibmmq helper

MarcialRosales committed:
Add extra logging around connection states

MarcialRosales committed:
Try to send capability in an array without any non-binary

MarcialRosales committed:
Skip test which sends illegal capability

MarcialRosales committed:
Try a bigger delay before starting to use ibm mq

MarcialRosales committed:
Test with non-binary capabilities

MarcialRosales committed:
Wait for connection and session events

MarcialRosales committed:
Log connection related events

added capture_logs function to the ibmmq_helper
but not called until I figure out how to add
extra logging. So far the only auth events captured
are those originated from the admin ui

MarcialRosales committed:
Add negative authentication test within ibmmq group

MarcialRosales committed:
Fix password

MarcialRosales committed:
Remove unnecessary command

MarcialRosales committed:
Bump up ibmmq docker image version

MarcialRosales committed:
Bump up ibmmq docker image

MarcialRosales committed:
Fix channel in test cases related to capabilities
